### PR TITLE
feature: add xi include loading on authoring initialization

### PR DESCRIPTION
**Pull request by Karol-Stelmaczonek**
Ticket: https://oat-sa.atlassian.net/browse/AUT-1647

## What's Changed

- Add loading of shared stimulus in text reader PCI in preview and authoring.

## Dependencies PRs

- https://github.com/oat-sa/extension-tao-test-preview-ui-loader/pull/166

## How to test
- Install & enable PCI Samples extension in Tao 
- Install & enable Tao Asset Manager extension
- Add shared stimulus via Asset Manager
- Go to items, create new item and add Text Reader custom interaction
- Add shared stimulus to text reader, click done.
- Click Preview and see that shared stimulus is visible.

## Video

https://github.com/oat-sa/extension-tao-itemqti/assets/118974926/733898a3-79f3-43f8-859e-4ee7afb1f96a

### CODE REVIEW

File: views/js/qtiCreator/helper/xincludeLoader.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                        });
+
+                        return Promise.all(xiIncludePromises).then(() => tempDiv.innerHTML);
+                    });
+
+                    return Promise.all(contentPromises).then(updatedContentItems => {
+                        page.content = updatedContentItems;
+                        return page;
+                    });
+                });
```
Comment (jsconan) [inline views/js/qtiCreator/helper/xincludeLoader.js:101]: **question:** Is this code specific to a particular PCI??? If yes, please make it generic, or move it to the PCI code

File: views/js/qtiCreator/itemCreator.js [hunk 2]
```diff
@@ -248,14 +250,26 @@ define([
...
                         _.forEach(item.getComposingElements(), function (element) {
                             if (element.is('customInteraction')) {
                                 usedCustomInteractionIds.push(element.typeIdentifier);
+
+                                if (element.properties && Array.isArray(element.properties.pages) && element.typeIdentifier === 'textReaderInteraction') {
+                                    const pagePromises = xincludeLoader.loadByElementPages(element.properties.pages);
+                                    elementPromises.push(Promise.all(pagePromises).then(updatedPages => {
+                                        element.properties.pages = updatedPages;
+                                    }));
+                                }
```
Comment (jsconan) [inline views/js/qtiCreator/itemCreator.js:264]: **quibble:** 😱 No, please not add code specific to one interaction that may even not exist... Specific code need to go to specific implementation. If needed, please add the proper adapter pattern.

**Review by KirylHatalski — APPROVED** [review]

**Review by jsconan — CHANGES_REQUESTED** [review]
While I understand the good intent, the implementation does not follow our standards... Please check comments below.

TL;DR: specific code must got to specific implementation. Please keep generic place only generic.

File: views/js/qtiCreator/widgets/interactions/states/Question.js [hunk 3]
```diff
@@ -23,10 +24,18 @@ define([
...
+        //render textReaderInteraction possible xi:include elements
+        let element = this.widget.element;
+        const baseUrl = this.widget.options.baseUrl;
+        if (element.properties && Array.isArray(element.properties.pages) && element.typeIdentifier === 'textReaderInteraction') {
+            const pagePromises = xincludeLoader.loadByElementPages(element.properties.pages, baseUrl)
+            Promise.all(pagePromises).then(updatedPages => {
+                element.properties.pages = updatedPages;
+                element.widgetRenderer.renderPages(element.properties);
+            });
+        }
```
Comment (jsconan) [inline views/js/qtiCreator/widgets/interactions/states/Question.js:38]: **quibble:** 😱 this is specific to a particular interaction, which is a PCI... it must be added to the Question state of the PCI, not to the state that any interaction will use... Please avoid spaghetti code.

File: views/js/qtiCreator/helper/xincludeLoader.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                        // require xml
+                        require([fileUrl], function (stimulusXml) {
+                            const data = self.parseXmlToDom(stimulusXml);
+                            resolve({ xinclude: xincludeElement, data }); // Resolve with an object containing xinclude and data
+                        }, function () {
+                            // In case the file does not exist, reject the promise
+                            reject(new Error('File not found'));
+                        });
+                    } else {
+                        reject(new Error('href or baseUrl is missing'));
```
Comment (viktar-dzmitryieu-tao) [inline views/js/qtiCreator/helper/xincludeLoader.js:45]: I'm getting this error when try to get back to the item with Stimulus in the Authoring

**Comment by viktar-dzmitryieu-tao** [comment]
From what I see current implementation does not corresponds to normal Stimulus functionality.
When we add the stimulus and save the item we inject all its content into the test-item, so we loose the connection to the stimulus (as I understand it should be just a link to the stimulus itself). I mean when we change the stimulus we should see the corresponding changes in the Item.
Apart from that we can change the stimulus content in the Item authoring, which should be possible only in stimulus authoring.
In authoring stimulus does not look like a separate monolith element and can't be removed at once.
